:books: translate chapter 7. Cost-Based Optimization

takaidohigasi opened the pull request:
## 原文
http://www.unofficialmysqlguide.com/cost-based-optimization.html

## see also
https://books.google.co.jp/books?id=SR_0DAAAQBAJ&pg=PT83&lpg=PT83&dq=mysql+explain+%E3%82%B3%E3%82%B9%E3%83%88&source=bl&ots=50sXYKlGGV&sig=co77davot2S73yzzoW-_wbrn5rs&hl=ja&sa=X&ved=0ahUKEwi9qb786NTSAhWKW7wKHWAVDaE4ChDoAQg_MAY#v=onepage&q=mysql%20explain%20%E3%82%B3%E3%82%B9%E3%83%88&f=false

## 備考

警告ブロックなどを表示するために_includesにテンプレート追加を実施しました。
デザインを考慮するのにコストがかかりそうだったため最小コストでtwitter bootstrap cssをincludeして標準のpanelでinfo, warningを使うようにしました。

takaidohigasi committed:
:books: translate chapter 7. cost based optimization

takaidohigasi committed:
:+1: add information and warning message box include template

takaidohigasi commented on _layouts/default.html:
```diff
@@ -4,7 +4,7 @@
     <meta http-equiv="X-UA-Compatible" content="chrome=1">
     <title>{{ site.title | default: site.github.repository_name }} by {{ site.github.owner_name }}</title>
-
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/css/bootstrap.min.css" media="screen">
```
先に定義したためクラスなどが重複するものは従来通りあとから読まれたものが優先される

takaidohigasi commented:
度々すみません...っ!!

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+
+![コストベース最適化](http://www.unofficialmysqlguide.com/_images/cost-based-optimization.png)
+
+上記が単純化であると述べたのは、オプティマイザは可能性のあるそれぞれの実行計画を徹底的に検索するわけではないからです。例えば、5つのテーブルが結合される場合、各テーブルに5つのインデックスがあるとすると、`5! * 5! =14400` 「を超える」クエリーの実行パターンがあることになります。
```
確かにoverは「を超える」かもしれませんが、日本語としてはちょっとかぎかっこの括り方に違和感があるので「超える」でいい気がします。

takaidohigasi replied:
ありがとうございます。そのように修正します！

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+
+上記が単純化であると述べたのは、オプティマイザは可能性のあるそれぞれの実行計画を徹底的に検索するわけではないからです。例えば、5つのテーブルが結合される場合、各テーブルに5つのインデックスがあるとすると、`5! * 5! =14400` 「を超える」クエリーの実行パターンがあることになります。
+
+* 各インデックスは1つ以上アクセス方法があります(例えば、インデックススキャン、レンジスキャンやインデックスルックアップです）。その上、各テーブルがテーブルスキャンを利用する可能性もあります。
```
かっこの始まりが半角で終わりが全角なのが気になってしまう私は細かすぎる！？

takaidohigasi replied:
気がついてませんでした... 確認して修正します

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+
+{% include info.html info="MySQL 8.0からメモリ内にあるインデックスの割合をコストモデルに利用する新機能を追加しています。以前のバージョンまでのMySQLではMySQLはIOにはページアクセスが常に必要であると仮定していました" %}
+
+コスト自体はリソース利用量を表す論理的な単位です。1単位にもはや正確な意味はありませんが、起源としては1990年代のハードディスクの1回のランダムIOと遡れます。
```
元々はそういう意味だったのか。。！(すいませんレビューと関係ないですw)

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+
+ハードウェアが進化したため、全ての構成要素でコストは一定とはみなせません(たとえばストレージのレイテンシーはSSDの出現と共に改善しました)。同様に、ソフトウェアもハードウェアの変化に対してとりくんでいるため(例えば、圧縮のような機能により)、リソース利用量も変化しうるでしょう。コンフィグ可能なコスト係数があれば、このようなケースに対して改善できます。
+
+{% include warning.html warn="コスト係数を修正すると多くのクエリーの計画は「悪くなる」可能性があることにご注意ください！ほとんどのプロダクション環境ではクエリーヒントを追加するのがよいでしょう。" %}
```
プロダクション環境、でも問題はない気がしますが、本番環境、の方がより一般的な気もします。

takaidohigasi replied:
プロダクション環境/商用環境/本番環境 → 本番環境としたいと思います！

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+
+{% include warning.html warn="コスト係数を修正すると多くのクエリーの計画は「悪くなる」可能性があることにご注意ください！ほとんどのプロダクション環境ではクエリーヒントを追加するのがよいでしょう。" %}
+
+{% include warning.html warn="次のエグザンプルに進む前に忘れずにコストを初期化するようにしてください。<br /><br />UPDATE mysql.server_cost SET cost_value=NULL WHERE cost_name='row_evaluate_cost';<br />
```
エグザンプル、は単に例で良いと思います(次の項で「例3で」と言っている)。

takaidohigasi replied:
Yes!

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+コスト自体はリソース利用量を表す論理的な単位です。1単位にもはや正確な意味はありませんが、起源としては1990年代のハードディスクの1回のランダムIOと遡れます。
+
+ハードウェアが進化したため、全ての構成要素でコストは一定とはみなせません(たとえばストレージのレイテンシーはSSDの出現と共に改善しました)。同様に、ソフトウェアもハードウェアの変化に対してとりくんでいるため(例えば、圧縮のような機能により)、リソース利用量も変化しうるでしょう。コンフィグ可能なコスト係数があれば、このようなケースに対して改善できます。
+
```
ここに入るはずの「Example 4 shows...」から始まる段落の訳が抜けている気がします。

takaidohigasi replied:
oh... sorry

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+{% include warning.html warn="次のエグザンプルに進む前に忘れずにコストを初期化するようにしてください。<br /><br />UPDATE mysql.server_cost SET cost_value=NULL WHERE cost_name='row_evaluate_cost';<br />
+FLUSH OPTIMIZER_COSTS;# セッションを閉じる" %}
+
+## メターデータと統計
```
メターデータ -> メタデータ

takaidohigasi replied:
失礼しました... ありがとうございます、修正します

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+
+| | インデックス情報 | ユニーク性 | Nullability |
+|:-|:-|:-|:-|
+| 概要 | ディクショナリはテーブル毎にインデックスの一覧の情報を持ちます | インデックスがユニークの場合、永久変換に利用され、実行計画のある部分を短縮するのに利用されます | オプティマイザはnullの可能性がある値を正しく扱う必要があります。列のNullabilityは一部の実行計画に影響を与えます | 
```
nullabilityって確かに英単語としてはありますが、「列がnullになり得るかどうか」といったように訳してしまった方が分かりやすいのではないでしょうか。

takaidohigasi replied:
そのようにします。
表のタイトルについてはNullabilityで残して表中の説明をNullability(列がnullになりうるかどうか)としたいと思います。

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+| 適用対象 | 全カラム | インデックスカラム | インデックスカラム |
+| 計算 | 事前 | 事前 | 利用するときに実施 |
+| 自動更新 | 通常オペレーション[^3]中 | テーブルの10%に変更が発生したとき| N/A |
+| 手動更新 | ANALYZE TABLE | ANALYZE TABLE | N/A |
```
ANALYZE TABLEはコマンドなので`ANALYZE TABLE`とした方がいいかもです。

takaidohigasi replied:
直してみます。表中のバッククォートがそのまま表示されてしまうようであればそのままにします。

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+| 概要 | テーブルの合計サイズの見積もりです | 少数のページ(デフォルト 20)をランダムでサンプルし、インデックスカラムの固有の値の数を推測します | オプティマイザがInnoDBに最小値と最大値を入力として問合せ、その範囲にある行数の見積もりを受け取ります |
+| 適用対象 | 全カラム | インデックスカラム | インデックスカラム |
+| 計算 | 事前 | 事前 | 利用するときに実施 |
+| 自動更新 | 通常オペレーション[^3]中 | テーブルの10%に変更が発生したとき| N/A |
```
N/Aは日本語の表ではあまり馴染みがないので、「-」で済ますか「適用なし」などとしてみたらどうでしょうか。

takaidohigasi replied:
そうかもしれません、

「-」にします

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+| 手動更新 | ANALYZE TABLE | ANALYZE TABLE | N/A |
+| 設定オプション | N/A | ページのサンプル数[^4] | 最大インデックスダイブ数[^5]、最大メモリ利用量[^6] |
+| 正確性 | 最も正確でない | データ分布の偏りに依存 | 最も正確 |
+| 通常の利用 | テーブルスキャンコストの決定に利用。インデックスがない場合の結合順に利用(大きいテーブルから先に) | 結合順の決定に利用。レンジの見積もりが最大インデックスダイブ数を超えた場合にも利用される | 述部の評価に利用。(利用可能なインデックスから何行マッチするかを推定)。population > 5000000がインデックスを使うべきでないかどうかを決めるのに利用する |
```
commonly used forは少し意訳して「使用頻度」とするのはどうでしょう。commonlyを通常と訳してしまうといまいちなんのことなのか分からないのではと思います。

takaidohigasi replied:
ありがとうございます。

commonlyは確かにわかりづらいですね。
しかし頻度とは少し違うように思います。

「一般的な使われ方」でどうでしょう

doublemarket replied:
あー、もう一度表を読み直して、そもそも表の意味自体を読み違えていたことに気づきました。一番左の列は、統計情報の特質を表しているのですね。となると、一般的な使われ方、がしっくりきますね。

読み違えに気づいて思ったのですが、例えば

- 計算 -> 計算のタイミング
- 自動更新 -> 自動更新のタイミング
- 手動更新 -> 手動更新の方法

などと左の列の記述を変えた方がより表として分かりやすくなりそうですね(横幅の問題がありそうですが。。)。

takaidohigasi replied:
ありがとうございます。

表を出してみたのですが現状でも結構厳しいのでコンパクトに記載した現状維持とさせてください!原文忠実ですので読み取れない範囲でもないのではないかな、と思います。
<img width="568" alt="2017-03-20 17 20 11" src="https://cloud.githubusercontent.com/assets/1700792/24092456/bb4c03c2-0d91-11e7-99f1-b21603a96c8f.png">

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+| 正確性 | 最も正確でない | データ分布の偏りに依存 | 最も正確 |
+| 通常の利用 | テーブルスキャンコストの決定に利用。インデックスがない場合の結合順に利用(大きいテーブルから先に) | 結合順の決定に利用。レンジの見積もりが最大インデックスダイブ数を超えた場合にも利用される | 述部の評価に利用。(利用可能なインデックスから何行マッチするかを推定)。population > 5000000がインデックスを使うべきでないかどうかを決めるのに利用する |
+
+{% include info.html info="統計を利用することで、品質見積もり時と商用環境で同じようにみえるクエリーが全く異なった実行となることがあります。クエリーの計画はデータ分布が変わるにつれて変化する可能性もあります" %}
```
in QA and productionは「品質保証のための環境と本番環境で」としてみてはどうでしょう。あるいは品質保証のための環境というより思い切ってステージング環境と書いてしまった方が、一般的なワークフローと合致して読者からすると分かりやすい気がします。

takaidohigasi replied:
ありがとうございます。ちょっと考えてます！！

takaidohigasi replied:
比較対象が「環境」とそろっていないので「品質保証のための環境と本番環境で」とそろえてみます。ステージング環境、については少し広義な意味として利用され、「性能テスト」「クエリーチューニング」などを実施することが想像しにくくなるため前者のままとしたいと思います。

doublemarket commented on _posts/2017-03-14-cost-based-optimization.md:
```diff
@@ -0,0 +1,83 @@
+| 正確性 | 最も正確でない | データ分布の偏りに依存 | 最も正確 |
+| 通常の利用 | テーブルスキャンコストの決定に利用。インデックスがない場合の結合順に利用(大きいテーブルから先に) | 結合順の決定に利用。レンジの見積もりが最大インデックスダイブ数を超えた場合にも利用される | 述部の評価に利用。(利用可能なインデックスから何行マッチするかを推定)。population > 5000000がインデックスを使うべきでないかどうかを決めるのに利用する |
+
+{% include info.html info="統計を利用することで、品質見積もり時と商用環境で同じようにみえるクエリーが全く異なった実行となることがあります。クエリーの計画はデータ分布が変わるにつれて変化する可能性もあります" %}
```
2つ目の文に「in production」の訳が含まれていないので、前の文との繋がりが少し分かりにくくなっていますね。「本番環境だと」とか文頭に入れた方がよさそうです。

takaidohigasi replied:
ありがとうございます、修正します!

doublemarket reviewed (changes requested):
遅くなりましたがレビューしましたー

takaidohigasi committed:
:books: improve chapter 7 translation

- fix bracket position
- fix mixed full-width/half-width bracket
- add back quote to the SQL
- modify 「N/A」 in the table to 「-」

takaidohigasi committed:
:books: add lacking translation for example 4 in chapter 7

takaidohigasi commented:
コメントいただいた点、修正しました。

翻訳の抜け
https://github.com/yakst/unofficialmysqlguide-ja/pull/37/commits/18d0e6bef88dccac41b756b81917a62ce7d4023f

その他コメント
https://github.com/yakst/unofficialmysqlguide-ja/pull/37/commits/2848f264850395a2abf9d0bbee3434fc61f552f2

にて反映しております！
